reputation/src/trs.rs

```diff
@@ -0,0 +1,747 @@
+        assert_eq!(a.get(&id1), Reputation(165));
+        assert_eq!(a.penalize(&id1, cnst(80)), Ok(Reputation(80)));
+        assert_eq!(a.get(&id1), Reputation(85));
+        // Check that the reputation was removed from the most recent "coin"
+        assert!(!a.queue.back().unwrap().1.contains_key(&id1));
+        // And the second most recent coin was mutated:
+        assert_eq!(
+            a.queue.iter().rev().nth(1).unwrap().1[&id1],
+            Reputation(15 - 10)
+        );
+    }
+
+    #[test]
+    fn penalize_all_coins() {
+        let mut a = Trs::new();
+        let id1 = "Alice".to_string();
+        assert_eq!(a.get(&id1), Reputation::default());
+        a.gain(Alpha(10), vec![(id1.clone(), Reputation(50))]);
+        a.gain(Alpha(11), vec![(id1.clone(), Reputation(30))]);
+        a.gain(Alpha(12), vec![(id1.clone(), Reputation(15))]);
+        a.gain(Alpha(13), vec![(id1.clone(), Reputation(70))]);
+        assert_eq!(a.get(&id1), Reputation(165));
+        assert_eq!(a.penalize(&id1, cnst(164)), Ok(Reputation(164)));
+        assert_eq!(a.get(&id1), Reputation(1));
+        // We only have 1 reputation from the first coin
+        assert_eq!(a.queue.front().unwrap().1[&id1], Reputation(1));
+    }
+
+    #[test]
+    fn penalize_and_expire() {
+        let mut a = Trs::new();
+        let id1 = "Alice".to_string();
+        assert_eq!(a.get(&id1), Reputation::default());
+        a.gain(Alpha(10), vec![(id1.clone(), Reputation(50))]);
+        a.gain(Alpha(11), vec![(id1.clone(), Reputation(30))]);
+        a.gain(Alpha(12), vec![(id1.clone(), Reputation(15))]);
+        a.gain(Alpha(13), vec![(id1.clone(), Reputation(70))]);
+        assert_eq!(a.get(&id1), Reputation(165));
+        assert_eq!(a.penalize(&id1, cnst(10)), Ok(Reputation(10)));
+        assert_eq!(a.expire(&Alpha(9)), Reputation(0));
+        assert_eq!(a.get(&id1), Reputation(155));
+        assert_eq!(a.penalize(&id1, cnst(10)), Ok(Reputation(10)));
+        assert_eq!(a.expire(&Alpha(10)), Reputation(50));
+        assert_eq!(a.get(&id1), Reputation(95));
+        assert_eq!(a.penalize(&id1, cnst(10)), Ok(Reputation(10)));
+        assert_eq!(a.expire(&Alpha(11)), Reputation(30));
+        assert_eq!(a.get(&id1), Reputation(55));
+        assert_eq!(a.penalize(&id1, cnst(10)), Ok(Reputation(10)));
+        assert_eq!(a.expire(&Alpha(12)), Reputation(15));
+        assert_eq!(a.get(&id1), Reputation(30));
+    }
+
+    #[test]
+    fn penalize_overflow() {
+        // An error is returned when the penalize function returns an invalid
+        // new reputation
+        let mut a = Trs::new();
+        let id1 = "Alice".to_string();
+        assert_eq!(a.get(&id1), Reputation::default());
+        a.gain(Alpha(10), vec![(id1.clone(), Reputation(50))]);
+        assert_eq!(a.get(&id1), Reputation(50));
+        // Penalize this identity with more reputation that it currently has
+        assert!(a.penalize(&id1, |_| Reputation(1000)).is_err());
```

Comment:
Why dont use this?
```suggestion
        assert!(a.penalize(&id1, const(1000)).is_err());
```

Reply:
The comment above is wrong, this tests for negative penalizations: an identity has 50 reputation and after penalization it will have 1000. This is impossible, so the penalize function returns an error.